app/server/appsmith-server/src/main/java/com/appsmith/server/services/GoogleRecaptchaServiceImpl.java

```
@@ -0,0 +1,79 @@
+package com.appsmith.server.services;
+
+import com.appsmith.server.configurations.GoogleRecaptchaConfig;
+import com.appsmith.server.exceptions.AppsmithError;
+import com.appsmith.server.exceptions.AppsmithException;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+import org.springframework.web.reactive.function.BodyInserters;
+import org.springframework.web.reactive.function.client.WebClient;
+import org.springframework.http.MediaType;
+import reactor.core.publisher.Mono;
+
+import java.net.URI;
+import java.util.HashMap;
+import java.util.Map;
+import java.time.Duration;
+
+@Service
+public class GoogleRecaptchaServiceImpl implements GoogleRecaptchaService {
+  private final WebClient webClient;
+
+  private final GoogleRecaptchaConfig googleRecaptchaConfig;
+
+  private static String BASE_URL = "https://www.google.com/recaptcha/api/";
+
+  private static String VERIFY_PATH = "/siteverify";
+
+  private final ObjectMapper objectMapper;
+  
+  private final Long timeoutInMillis = Long.valueOf(10000);
+
+  @Autowired
+  public GoogleRecaptchaServiceImpl(WebClient.Builder webClientBuilder,
+                                    GoogleRecaptchaConfig googleRecaptchaConfig, ObjectMapper objectMapper) {
+      this.webClient = webClientBuilder.baseUrl(BASE_URL).build();
+      this.googleRecaptchaConfig = googleRecaptchaConfig;
+      this.objectMapper = objectMapper;
+  }
+
+  @Override
+  public Mono<Boolean> verify(String recaptchaResponse){
+    
+    // if not enabled or secret key is not configured, abort verification.
+    if (!googleRecaptchaConfig.isEnabled() || googleRecaptchaConfig.getSecretKey() == "") {
```

Question: Should we raise an error in case `enabled` flag is set to `true` but `secretKey` is not?

---

Is there a reason we need the `isEnabled` flag? Can we simply check the existence of the `secretKey` to check if Recaptcha is available? If the key is not present, we assume that recaptcha feature is disabled.

---

Reason is to make it explicit, handling based on secretKey might lead to confusion and one needs to dive into code to understand. Documenting behavioir based on `secretKey` should help.

---

Hmm. I’m generally of the opinion that minimal configuration is more beneficial. Also having to configure  multiple keys for a single feature seems counter intuitive to me. 

Hence I’m wondering if the Enabled flag is relevant here. Is there any scenario where the flag is enabled but the key is not set? If not, then one of them is a redundant parameter

---

Understood, make sense.

---

Updated it.